Figaro/src/main/scala/com/cra/figaro/algorithm/factored/factors/factory/ChainFactory.scala

```diff
@@ -62,5 +71,44 @@ object ChainFactory {
     })
     pairFactor :: tempFactors
   }
-  
+
+  def makeSingleFactor[T, U](cc: ComponentCollection, chain: Chain[T, U])(implicit mapper: PointMapper[U]): List[Factor[Double]] = {
+    val chainComp = cc(chain)
+    val parentVar = Factory.getVariable(cc, chain.parent)
+    val childVar = Factory.getVariable(cc, chain)
+    val factor = new BasicFactor[Double](List(parentVar), List(childVar))
+    for { parentIndex <- 0 until parentVar.range.length } {
+      val parentXV = parentVar.range(parentIndex)
+      if (parentXV.isRegular && chainComp.subproblems.contains(parentXV.value) && !chainComp.subproblems(parentXV.value).solution.isEmpty) {
+        val subproblem = chainComp.subproblems(parentXV.value)
+        // Need to normalize subsolution in case there's any nested evidence
+        val subsolution = subproblem.solution.reduce(_.product(_))
+        val sum = subsolution.foldLeft(subsolution.semiring.zero, subsolution.semiring.sum(_, _))
+        val subVars = subsolution.variables
+        if (subVars.length == 1) {
+          val subVar = subVars(0)
+          for { subVal <- subVar.range } {
+            val childIndex = childVar.range.indexOf(subVal)
+            val subIndex = subVar.range.indexOf(subVal)
+            val entry = subsolution.semiring.product(subsolution.get(List(subIndex)), 1.0 / sum)
```

Comment:
This is not semiring safe. The 1.0 should be replaced by semiring.one, and the divide should be done through a semiring operation. However, this is only valid for divideable semirings. @apfeffer please review this.